### PR TITLE
feat(chat): persist prompt history to ~/.amaebi/history.jsonl, filter by cwd

### PR DESCRIPTION
**Pull request by yuankuns**
## Motivation

`amaebi chat`'s prompt history has always been in-memory only — exit
the client and the ↑ buffer is gone.  User expected Claude Code's
behavior: history persists across sessions, filtered by project
directory.

(Scope note: `amaebi ask` uses a plain `BufReader` on stdin — not
rustyline — so its prompt is intentionally NOT persisted by this PR.
Extending persistence to `ask` would require threading rustyline into
that path and is out of scope here.  `docs/chat.md` spells this out.)

## Scope

- New append-only JSONL at `~/.amaebi/history.jsonl`, schema mirrors
  Claude Code's (`display / pasted_contents / timestamp_ms / cwd /
  session_id`).
- Every successful `read_line_raw` (chat prompt only) appends one row.
- First readline on a TTY lazily loads the file, filters rows by the
  canonicalised cwd (via `session::canonical_key`, so symlinks /
  bind-mounts / `..` variants all match), and seeds rustyline's
  in-memory history.
- File created with mode 0o600 (matches `sessions.json` style); mode
  is re-applied on existing files whose permissions are looser.
- No explicit locking — relies on typical Linux + local-FS behaviour
  that a single `write(2)` under `MAX_LINE_BYTES` to an `O_APPEND`
  regular file is not interleaved with concurrent `O_APPEND` writers.
  Not a POSIX guarantee; documented as best-effort in the code.
- Oversize rows are pre-truncated field-by-field (display shrinks by
  halves), keeping the output JSON always well-formed.  If even the
  fixed fields overflow the budget, the row is silently dropped
  rather than emitted oversize.
- Loader reads with a bounded per-line buffer (fill_buf/consume
  chunking) so a hostile/corrupted oversize line cannot OOM the
  loader.
- Single `write(2)` per row with short-write treated as error, so a
  retry loop never splits a record mid-line.
- `docs/chat.md` updated with a short note on the new behavior.
- 5 new unit tests (round-trip, cwd filter, corrupted-line tolerance,
  oversize truncation, missing-file-is-ok).

## Manual e2e

1. `cargo build --release` from this branch; restart daemon.
2. In a terminal at `/home/yuankuns/libraries.ai.cutlass.internal`, run
   `amaebi chat`; type `/model` then Enter; type `hello` then Enter;
   exit chat.
3. `cat ~/.amaebi/history.jsonl` — should show 2 lines with
   `cwd: "/home/yuankuns/libraries.ai.cutlass.internal"` (canonical
   form) and the expected `display` values.
4. Restart `amaebi chat` in the SAME directory; press ↑ — `hello`
   appears.  Press ↑ again — `/model` appears.
5. `cd /home/yuankuns/amaebi && amaebi chat` — press ↑ — previous
   project's commands do NOT appear.  (Confirms cwd filter.)
6. From two terminals in the SAME cwd, start two chats simultaneously,
   alternate typing commands in each — both histories persist, no
   corruption on inspection of `history.jsonl`.

## Rollback

`rm ~/.amaebi/history.jsonl` + revert; no data migration needed.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/client.rs [hunk 1]
```diff
@@ -2302,13 +2345,176 @@ mod prompt_input {
...
+        json.push('\n');
+        json
+    }
+
+    /// Append one line of prompt history to `~/.amaebi/history.jsonl`,
+    /// creating the file on first use.  POSIX atomicity: a `write(2)`
+    /// smaller than `PIPE_BUF` (4096 on Linux) to an `O_APPEND` file is
+    /// atomic with respect to other `O_APPEND` writers, so two amaebi
+    /// clients in the same cwd can safely race without corrupting each
+    /// other's rows.  [`build_history_line`] enforces that ceiling.
```
[the remark this answers is not in the document]
Comment (yuankuns) [reply src/client.rs]: Addressed in an earlier round: `append_history_line_to_path` now issues a single `f.write(buf)` (not `write_all`) and treats any short write as a hard error rather than retrying. See src/client.rs:2598 and the doc comment just above. An `flock` still isn't used — documented as best-effort.

File: src/client.rs [hunk 2]
```diff
@@ -2302,13 +2345,176 @@ mod prompt_input {
...
+        append_history_line_to_path(&path, &json)
+    }
+
+    /// Raw append to `path`.  Split out so tests can target a tempfile.
+    fn append_history_line_to_path(path: &Path, json_with_newline: &str) -> std::io::Result<()> {
+        use std::io::Write as _;
+        let mut f = std::fs::OpenOptions::new()
+            .create(true)
+            .append(true)
+            .open(path)?;
```
[the remark this answers is not in the document]
Comment (yuankuns) [reply src/client.rs]: Addressed in an earlier round: file is opened with `OpenOptionsExt::mode(0o600)` and, if an existing file's mode is looser, we call `set_permissions` to bring it back to 0o600. See src/client.rs:2578-2593.

File: src/client.rs [hunk 3]
```diff
@@ -2302,13 +2345,176 @@ mod prompt_input {
...
+    /// atomic with respect to other `O_APPEND` writers, so two amaebi
+    /// clients in the same cwd can safely race without corrupting each
+    /// other's rows.  [`build_history_line`] enforces that ceiling.
+    fn append_history_line(display: &str) -> std::io::Result<()> {
+        let path = default_history_path()?;
+        if let Some(parent) = path.parent() {
+            std::fs::create_dir_all(parent)?;
+        }
+        let cwd = std::env::current_dir()
+            .map(|p| p.to_string_lossy().into_owned())
```
[the remark this answers is not in the document]
Comment (yuankuns) [reply src/client.rs]: Addressed in an earlier round: `cwd_for_history` now canonicalises via `session::canonical_key`, matching the rest of the codebase. Both writer and loader go through this helper. See src/client.rs:2537-2542.

File: src/client.rs [hunk 4]
```diff
@@ -2302,13 +2345,176 @@ mod prompt_input {
...
+    /// Load `~/.amaebi/history.jsonl`, filter rows by the current process
+    /// cwd, and feed the matches into `editor`'s in-memory history in
+    /// file order (oldest first).  Malformed / mid-write lines are
+    /// skipped rather than aborting the load.
+    fn load_history_for_current_cwd(editor: &mut rustyline::DefaultEditor) -> std::io::Result<()> {
+        let path = default_history_path()?;
+        let cwd = std::env::current_dir()
+            .map(|p| p.to_string_lossy().into_owned())
+            .unwrap_or_default();
+        load_history_from_path(&path, &cwd, editor)
```
[the remark this answers is not in the document]
Comment (yuankuns) [reply src/client.rs]: Addressed in an earlier round: both the writer and the loader go through `cwd_for_history` -> `session::canonical_key`, so symlink / bind-mount / `..` variants of the same project match. See src/client.rs:2537-2542 (shared helper) and 2660 (used by the loader).

File: src/client.rs [hunk 5]
```diff
@@ -2302,13 +2345,176 @@ mod prompt_input {
...
+            Ok(f) => f,
+            Err(e) if e.kind() == std::io::ErrorKind::NotFound => return Ok(()),
+            Err(e) => return Err(e),
+        };
+        use std::io::BufRead as _;
+        let reader = std::io::BufReader::new(file);
+        for line in reader.lines() {
+            let Ok(line) = line else { continue };
+            let Ok(row) = serde_json::from_str::<HistoryRow>(&line) else {
+                continue;
```
[the remark this answers is not in the document]
Comment (yuankuns) [reply src/client.rs]: Addressed in an earlier round: replaced `BufRead::lines()` with a bounded `fill_buf()` + `consume()` scan capped at `READ_CAP = MAX_LINE_BYTES + 64`. Lines exceeding the cap are skipped (bytes discarded) until the next newline without ever materialising into a large `String`. See src/client.rs:2697-2780.

File: docs/chat.md [hunk 6]
```diff
@@ -116,6 +116,16 @@ The prompt input uses `rustyline` for line editing — arrow keys, Emacs
 bindings, CJK-aware cursor movement, and persistent history across
 sessions. This replaced a hand-rolled editor; see commit `0bb236a`.
 
+Every successful prompt is appended to `~/.amaebi/history.jsonl` (one
+JSON object per line — `display`, `pasted_contents`, `timestamp_ms`,
+`cwd`, `session_id`). When `amaebi chat` / `amaebi ask` opens, it seeds
+the in-memory ring with rows whose `cwd` matches the current working
+directory, so pressing ↑ in a given project shows prior commands from
```
[the remark this answers is not in the document]
Comment (yuankuns) [reply docs/chat.md]: Addressed in an earlier round: docs/chat.md now explicitly scopes the persistence claim to `amaebi chat`'s interactive prompt and calls out that `amaebi ask` uses a plain `BufReader` on stdin, which is not persisted. See docs/chat.md:119-136. PR description also updated to match.

File: src/client.rs [hunk 7]
```diff
@@ -2302,13 +2345,263 @@ mod prompt_input {
...
+        // Cap on how much we allow a single line to consume before we
+        // give up on it — slightly above MAX_LINE_BYTES so a legitimate
+        // written-at-the-cap line still parses while anything bigger is
+        // treated as garbage and skipped without OOMing the loader.
+        const READ_CAP: usize = MAX_LINE_BYTES + 64;
+        let mut buf: Vec<u8> = Vec::with_capacity(512);
+        loop {
+            buf.clear();
+            let n = reader.read_until(b'\n', &mut buf)?;
+            if n == 0 {
```
[the remark this answers is not in the document]
Comment (yuankuns) [reply src/client.rs]: Addressed in an earlier round: we no longer use `read_until`. The loader now scans `fill_buf()` for `\\n` and consumes in chunks; the in-memory line buffer never grows past `READ_CAP = MAX_LINE_BYTES + 64` even on a pathological single-line file. See src/client.rs:2697-2780.

File: src/client.rs [hunk 8]
```diff
@@ -2302,13 +2345,263 @@ mod prompt_input {
...
+            // tolerates parse errors anyway).
+            const MARKER: &str = "…(truncated)\"}";
+            let target = MAX_LINE_BYTES.saturating_sub(MARKER.len() + 1);
+            let mut cut = target.min(json.len());
+            while cut > 0 && !json.is_char_boundary(cut) {
+                cut -= 1;
+            }
+            json.truncate(cut);
+            json.push_str(MARKER);
+        }
```
[the remark this answers is not in the document]
Comment (yuankuns) [reply src/client.rs:2547]: Addressed in an earlier round: `build_history_line` now truncates `display` and `pasted_contents` at the field level BEFORE serialisation, so the emitted JSON is always well-formed. The shrink loop halves `display` and re-serialises each iteration instead of slicing raw bytes out of the final JSON string. See src/client.rs:2485-2525.

File: src/client.rs [hunk 9]
```diff
@@ -2185,6 +2187,39 @@ mod prompt_input {
...
+    /// an `O_APPEND` file is atomic with respect to other `O_APPEND`
+    /// writers on the same file (Linux: `PIPE_BUF = 4096`).  Staying well
+    /// under that ceiling keeps our "no explicit lock" design safe even
+    /// when two `amaebi chat` processes in the same cwd write at once.
+    const MAX_LINE_BYTES: usize = 3800;
+
+    /// Maximum byte length retained for `pasted_contents` before we
+    /// truncate with a marker.  Not strictly necessary today — we do not
+    /// currently populate `pasted_contents` — but the schema reserves the
+    /// field and the truncation keeps future writers inside `PIPE_BUF`.
```
[the remark this answers is not in the document]
Comment (yuankuns) [reply src/client.rs]: Addressed in an earlier round: the comment now explicitly says this is a Linux + local-FS best-effort sizing choice, NOT a POSIX atomicity guarantee for regular files, and notes that strict inter-process record atomicity would require an explicit lock. See src/client.rs:2194-2215.

File: src/client.rs [hunk 10]
```diff
@@ -2185,6 +2187,39 @@ mod prompt_input {
...
+    const MAX_PASTED_BYTES: usize = 200;
+
+    /// One line of the on-disk prompt history.  Mirrors Claude Code's
+    /// `~/.claude/history.jsonl` schema so the file is human-greppable and
+    /// the semantics are familiar.  `cwd` is the `std::env::current_dir()`
+    /// of the process that captured the line; the loader filters by it so
+    /// ↑ inside a project only surfaces that project's prompts.
```
[the remark this answers is not in the document]
Comment (yuankuns) [reply src/client.rs]: Addressed in an earlier round: `HistoryRow` doc now says the persisted `cwd` is a canonicalised path string derived from the capturing process's current directory, and notes the filter is stable across symlink / `..` variants. See src/client.rs:2218-2234.

File: docs/chat.md [hunk 11]
```diff
@@ -116,6 +116,24 @@ The prompt input uses `rustyline` for line editing — arrow keys, Emacs
...
+intentional: a chat started in `~/projectA` does not replay prompts
+from `~/projectB`. Canonicalisation handles symlink / `..` / bind-mount
+variants of the same directory. To see every line across every
+project, `grep ~/.amaebi/history.jsonl`.
+
+Note: only `amaebi chat`'s interactive prompt feeds this file.
+`amaebi ask` reads its single prompt + any steer-correction text via a
+plain `BufReader` on stdin (no rustyline), so those lines are not
+persisted and are not recalled by ↑ the next time you run
+`amaebi chat`.
```
[the remark this answers is not in the document]
Comment (yuankuns) [reply docs/chat.md:136]: Intentional scope: this PR is chat-only. `amaebi ask` reads its prompt via a plain `BufReader` on stdin (not rustyline), so extending persistence there would require a larger rewrite of the ask-side input path. The PR description's first paragraph now calls this out explicitly, and docs/chat.md:132-136 spells it out. Tracking that as a possible follow-up rather than expanding this PR.

File: src/client.rs [hunk 12]
```diff
@@ -2302,13 +2353,329 @@ mod prompt_input {
...
+                        // would overflow the cap.
+                        let remaining = READ_CAP.saturating_sub(buf.len());
+                        let copy_bytes = remaining.min(idx);
+                        buf.extend_from_slice(&chunk[..copy_bytes]);
+                        if copy_bytes == idx {
+                            try_load_row(&buf, cwd, editor);
+                        }
+                        // Else: we hit the cap mid-line; the rest of
+                        // the line is already past the newline, which
+                        // we never saw — fall-through to reset.
```
[the remark this answers is not in the document]
Comment (yuankuns) [reply src/client.rs]: This comment seems to have landed on the wrapper `load_history_for_current_cwd` at src/client.rs:2678 — but the substantive bounded-read concern it's adjacent to was addressed in earlier rounds: the inner `load_history_from_path` uses a bounded `fill_buf()`/`consume()` scan, and we now additionally seek to the tail of files larger than `LOAD_TAIL_BYTES` so startup I/O is bounded independent of history size. See src/client.rs:2700-2722 in the new commit.

File: src/client.rs [hunk 13]
```diff
@@ -2302,13 +2353,329 @@ mod prompt_input {
...
+        if s.len() <= max {
+            return s.to_owned();
+        }
+        let mut cut = max;
+        while cut > 0 && !s.is_char_boundary(cut) {
+            cut -= 1;
+        }
+        let mut out = String::with_capacity(cut + 12);
+        out.push_str(&s[..cut]);
+        out.push_str("…(truncated)");
```
[the remark this answers is not in the document]
Comment (yuankuns) [reply src/client.rs]: Addressed in an earlier round: `truncate_utf8_with_marker` is now monotonic — when `max < MARKER.len()` it returns "" (empty string) rather than appending the marker, so repeated calls with smaller `max` strictly shrink. This guarantees the `build_history_line` shrink loop terminates. See src/client.rs:2460-2478 for the implementation and the "Degenerate case" doc block just above it.

File: src/client.rs [hunk 14]
```diff
@@ -2302,13 +2353,329 @@ mod prompt_input {
...
+            capped_display = truncate_utf8_with_marker(&capped_display, target.max(1));
+            json = serialize(&capped_display, &capped_pasted);
+            // Degenerate guard: if halving stops making progress (e.g.
+            // pasted_contents alone blows the budget), break rather
+            // than spin forever.  The load path tolerates a dropped row.
+            if capped_display.len() < 8 {
+                break;
+            }
+        }
+
```
[the remark this answers is not in the document]
Comment (yuankuns) [reply src/client.rs:2548]: Addressed in an earlier round: after the shrink loop, `build_history_line` has a final size check that returns "" (drop the row) if the line still exceeds `MAX_LINE_BYTES` — the caller (`append_history_line_to_path`) no-ops on empty input. See src/client.rs:2535-2545 for the drop logic.

File: src/client.rs [hunk 15]
```diff
@@ -2302,13 +2353,370 @@ mod prompt_input {
...
+        // When there's no room for the marker + at least one byte of
+        // content, fall back to a marker-only (or marker-prefix) string
+        // clipped to `max`.  Preserves the "output ≤ max" invariant
+        // even in degenerate cases where the caller demands a tiny cap.
+        if max <= MARKER.len() {
+            let mut cut = max;
+            while cut > 0 && !MARKER.is_char_boundary(cut) {
+                cut -= 1;
+            }
+            return MARKER[..cut].to_owned();
```
[the remark this answers is not in the document]
Comment (yuankuns) [reply src/client.rs]: The empty return is intentional and documented at src/client.rs:2451-2458 ("Degenerate case... the function returns the empty string"). Real callers only hit this path when the input is already degenerate (a display whose budget was halved below 14 bytes means the cwd/session_id fields already dominate the budget). The new commit (17083d7) turns that empty-display case into a full row drop in `build_history_line`, so a degenerate prompt is no longer persisted as a blank `display` entry — see fix #3 in that commit and the `build_history_line_drops_row_when_display_truncates_to_empty` test at src/client.rs:3058.

File: src/client.rs [hunk 16]
```diff
@@ -2302,13 +2353,372 @@ mod prompt_input {
...
+        opts.create(true).append(true);
+        #[cfg(unix)]
+        opts.mode(0o600);
+        let mut f = opts.open(path)?;
+
+        #[cfg(unix)]
+        {
+            let current = f.metadata()?.permissions().mode() & 0o777;
+            if current != 0o600 {
+                let mut perms = f.metadata()?.permissions();
```
[the remark this answers is not in the document]
Comment (yuankuns) [reply src/client.rs]: Addressed in an earlier round: we now call `f.metadata()?` once, read `perms` from that handle, check the mode, and only call `set_permissions` when needed. See src/client.rs:2653-2661.

File: docs/chat.md [hunk 17]
```diff
@@ -116,6 +116,24 @@ The prompt input uses `rustyline` for line editing — arrow keys, Emacs
...
+to `~/.amaebi/history.jsonl` (one JSON object per line — `display`,
+`pasted_contents`, `timestamp_ms`, `cwd`, `session_id`). When
+`amaebi chat` opens, it seeds the in-memory ring with rows whose
+canonicalised `cwd` matches the current working directory, so pressing
+↑ in a given project shows prior commands from that same project
+(including slash commands like `/claude` and `/model`). The filter is
+intentional: a chat started in `~/projectA` does not replay prompts
+from `~/projectB`. Canonicalisation handles symlink / `..` / bind-mount
+variants of the same directory. To see every line across every
+project, `grep ~/.amaebi/history.jsonl`.
```
[the remark this answers is not in the document]
Comment (yuankuns) [reply docs/chat.md]: Addressed in an earlier round: docs/chat.md:129-130 now shows `cat ~/.amaebi/history.jsonl` for inspection and `grep -F /claude ~/.amaebi/history.jsonl` as a valid search example.

File: src/client.rs [hunk 18]
```diff
@@ -2302,13 +2353,372 @@ mod prompt_input {
                 // so Up/Down navigation skips accidental bare-Enter presses.
                 if !line.is_empty() {
                     let _ = ed.add_history_entry(line.as_str());
+                    // Persist to disk for cross-session recall.  Best-effort:
+                    // an I/O failure here is not worth surfacing to the user
+                    // — log at debug and continue.  See `append_history_line`
+                    // for atomicity notes.
+                    if let Err(e) = append_history_line(&line) {
+                        tracing::debug!(error = %e, "failed to append to history.jsonl");
+                    }
```
[the remark this answers is not in the document]
Comment (yuankuns) [reply src/client.rs:2418]: Addressed in an earlier round: the `ed` guard is dropped before `append_history_line` runs, so a slow disk can't block a concurrent `spawn_blocking(read_line_raw)` from acquiring the editor lock. See src/client.rs:2421-2440.

File: src/client.rs [hunk 19]
```diff
@@ -2302,13 +2353,372 @@ mod prompt_input {
...
+    ///    (mirror `~/.amaebi/sessions.json` and similar state files)
+    ///    and defensively re-apply 0o600 on an existing file whose
+    ///    permissions are looser — covers the case where umask on a
+    ///    prior run was wrong or someone changed the mode manually.
+    /// 2. **Single `write(2)` per record.** `O_APPEND` + one
+    ///    `write(2)` ≤ `PIPE_BUF` is atomic across processes.
+    ///    `std::io::Write::write_all` will loop on short writes which
+    ///    can interleave concurrent appends; we call the low-level
+    ///    `write` once and treat a short write as a hard error so the
+    ///    caller doesn't retry and split the record.
```
[the remark this answers is not in the document]
Comment (yuankuns) [reply src/client.rs]: Addressed in an earlier round: the doc comment on `append_history_line_to_path` now explicitly says single-`write(2)` O_APPEND delivery is Linux + local-FS best-effort, NOT a POSIX guarantee for regular files, and that strict atomicity would require an explicit lock. Language matches the `MAX_LINE_BYTES` constant comment. See src/client.rs:2612-2628.

File: src/client.rs [hunk 20]
```diff
@@ -2286,27 +2329,416 @@ mod prompt_input {
...
+    /// newline can never force the in-memory line buffer above
+    /// [`READ_CAP`].  Lines whose length hits [`READ_CAP`] are skipped
+    /// entirely — we keep consuming bytes (discarded) until we find the
+    /// next `\n` and then resume parsing, so one bad row doesn't lose
+    /// the rest of the file.
+    fn load_history_from_path(
+        path: &Path,
+        cwd: &str,
+        editor: &mut rustyline::DefaultEditor,
+    ) -> std::io::Result<()> {
```
[the remark this answers is not in the document]
Comment (yuankuns) [reply src/client.rs:2701]: Fixed in 17083d7: `load_history_from_path` now seeks to `len - LOAD_TAIL_BYTES` (4 MiB) on oversized files and uses the existing `discarding` state to drop the partial head of the first surviving line, preserving oldest-first order for the rest. New test `load_history_bounds_tail_for_oversized_file` at src/client.rs:3020 writes >4 MiB and asserts the tail marker is the last loaded row and total count < input.

File: src/client.rs [hunk 21]
```diff
@@ -2286,27 +2329,416 @@ mod prompt_input {
...
+    /// [`append_history_line_to_path`] issues exactly one `write(2)`
+    /// syscall (no retry loop) so a short write surfaces as an error
+    /// rather than silently producing a split/interleaved record.
+    fn append_history_line(display: &str) -> std::io::Result<()> {
+        let path = default_history_path()?;
+        if let Some(parent) = path.parent() {
+            std::fs::create_dir_all(parent)?;
+        }
+        let cwd = cwd_for_history();
+        let json = build_history_line(display, "", &cwd, "", now_ms());
```
[the remark this answers is not in the document]
Comment (yuankuns) [reply src/client.rs]: Fixed in 17083d7: added `prompt_input::SESSION_ID: OnceLock<String>` with a `set_session_id(&session_id)` call from `run_chat_loop` right after the session id is resolved (src/client.rs:1004-1008). `append_history_line` now reads this via `current_session_id()` and threads it into `build_history_line`, so persisted rows carry the actual chat UUID. Chose a module-level set-once slot rather than a `read_line_raw` parameter because the three call sites pass the bare fn pointer to `spawn_blocking`, and adding an arg would require closure-wrapping at every site. New test `build_history_line_includes_session_id_in_output` at src/client.rs:3085 pins this.

File: src/client.rs [hunk 22]
```diff
@@ -2286,27 +2329,416 @@ mod prompt_input {
...
+        }
+        let content_budget = max - MARKER.len();
+        let mut cut = content_budget;
+        while cut > 0 && !s.is_char_boundary(cut) {
+            cut -= 1;
+        }
+        let mut out = String::with_capacity(cut + MARKER.len());
+        out.push_str(&s[..cut]);
+        out.push_str(MARKER);
+        out
```
[the remark this answers is not in the document]
Comment (yuankuns) [reply src/client.rs:2480]: Fixed in 17083d7: `build_history_line` now drops the row entirely (returns "", caller no-ops) when a non-empty input `display` would be persisted as an empty string after truncation. The existing 'total oversize' bail-out could miss this case because the serialised row can still fit once `display` is empty. New test `build_history_line_drops_row_when_display_truncates_to_empty` at src/client.rs:3058 constructs the exact edge case (cwd_size=3715) to verify the new drop fires.

File: src/client.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (yuankuns) [reply src/client.rs]: Addressed in the commit after this review round: the docstring at src/client.rs:2210-2219 now explicitly calls out that POSIX's `PIPE_BUF` atomicity applies to pipes/FIFOs NOT regular files, and that MAX_LINE_BYTES is a best-effort Linux+local-FS sizing budget rather than a portable guarantee. An explicit `flock` is intentionally declined for the user-convenience history.

**Review by yuankun — APPROVED** [review]

File: src/client.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (yuankuns) [reply src/client.rs]: Addressed in the commit after this review round: HistoryRow's docstring at src/client.rs:2229-2235 now says `cwd` stores a *canonicalized* path via `session::canonical_key`, and both writer (`cwd_for_history` at ~2590) and loader share that helper so symlink / `..` / bind-mount path variants all match.

File: docs/chat.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (yuankuns) [reply docs/chat.md:136]: Addressed in the commit after this review round: docs/chat.md:132-136 scopes the persistence claim to `amaebi chat`'s interactive (rustyline) prompt and explicitly notes that `amaebi ask` uses a plain BufReader on stdin with no persistence. PR description updated to match.

File: src/client.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (yuankuns) [reply src/client.rs]: Addressed in the commit after this review round: the comment at src/client.rs:2757-2761 now says "this line exceeded READ_CAP; we filled `buf` up to the cap and now discard the rest of the oversize line up to the newline we just found, then fall through to reset and resume parsing at the next line boundary" — accurately describing the found-newline oversize path.

File: src/client.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (yuankuns) [reply src/client.rs]: Addressed in the commit after this review round: `truncate_utf8_with_marker` at src/client.rs:2463-2481 now returns `String::new()` when `max < MARKER.len()` instead of appending the marker, making it strictly monotonic (output length ≤ requested max). The docstring at 2449-2462 calls out this behaviour and the callers' safety. The shrink loop in `build_history_line` also keeps a `last_display_len` guard that bails if progress stops, so infinite-loop is impossible.

File: src/client.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (yuankuns) [reply src/client.rs:2548]: Addressed in the commit after this review round: `build_history_line` has an explicit post-shrink guard at src/client.rs:2545-2547 that returns an empty string when `json.len() + 1 > MAX_LINE_BYTES`; the caller (`append_history_line_to_path` at 2626-2628) short-circuits on the empty string and writes nothing, so an oversize line can never be emitted.

File: src/client.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (yuankuns) [reply src/client.rs]: Addressed in the commit after this review round: `truncate_utf8_with_marker` no longer tries to emit a marker-prefix under tight caps — when `max < MARKER.len()` it returns `String::new()` and `build_history_line`'s caller side drops the row (see src/client.rs:2556-2558 and 2545). Docstring at 2454-2462 documents this explicitly.

File: src/client.rs (hunk 16 above)
[the remark this answers is not in the document]
Comment (yuankuns) [reply src/client.rs]: Addressed in the commit after this review round: src/client.rs:2640-2648 now reads metadata once (`let mut perms = f.metadata()?.permissions();`) and reuses that `Permissions` handle for both the mode check and the `set_permissions` call. Inline comment at 2641-2643 documents the single-syscall intent.

File: docs/chat.md (hunk 17 above)
[the remark this answers is not in the document]
Comment (yuankuns) [reply docs/chat.md]: Addressed in the commit after this review round: docs/chat.md:128-130 now says "to inspect every line across every project use `cat ~/.amaebi/history.jsonl`; to search, pass a pattern (e.g. `grep -F /claude ~/.amaebi/history.jsonl`)" — both invocations are valid.

File: src/client.rs (hunk 18 above)
[the remark this answers is not in the document]
Comment (yuankuns) [reply src/client.rs:2418]: Addressed in the commit after this review round: `read_line_raw` at src/client.rs:2402-2419 now `drop(ed)` BEFORE calling `append_history_line`, so the global EDITOR mutex is never held across the filesystem write. Comment at 2402-2409 documents the invariant.

File: src/client.rs (hunk 19 above)
[the remark this answers is not in the document]
Comment (yuankuns) [reply src/client.rs]: Addressed in the commit after this review round: the `append_history_line_to_path` doc at src/client.rs:2609-2618 now explicitly says "This is NOT a POSIX guarantee for regular files — POSIX defines `PIPE_BUF` atomicity only for pipes/FIFOs — so treat it as best-effort." Wording matches the earlier caveat on `MAX_LINE_BYTES`.

File: src/client.rs (hunk 20 above)
[the remark this answers is not in the document]
Comment (yuankuns) [reply src/client.rs:2701]: Addressed in the commit after this review round: `load_history_from_path` at src/client.rs:2721-2727 seeks to `file_len - LOAD_TAIL_BYTES` for files larger than 4 MiB and discards the partial first line, so startup I/O is bounded to O(LOAD_TAIL_BYTES) regardless of file size. Docstring at 2692-2708 documents the bound. Compaction/ring-buffer isn't needed because the tail window already keeps HISTORY_CAP=1000 rows worth of matches.

File: src/client.rs (hunk 21 above)
[the remark this answers is not in the document]
Comment (yuankuns) [reply src/client.rs]: Addressed in the commit after this review round: `set_session_id` in `run_chat_loop` (src/client.rs:~1004) threads the active chat UUID into a module-level `SESSION_ID`, and `append_history_line` reads it via `current_session_id()` so every on-disk row now carries its real session_id. Test `build_history_line_includes_session_id_in_output` guards the schema.

File: src/client.rs (hunk 22 above)
[the remark this answers is not in the document]
Comment (yuankuns) [reply src/client.rs:2480]: Addressed in the commit after this review round: `build_history_line` at src/client.rs:2549-2558 drops the row (returns `""`) when a non-empty input truncates to an empty display. The caller then writes nothing, so no blank-display row ever lands on disk. Regression test `build_history_line_drops_row_when_display_truncates_to_empty` guards this.
